smarts/core/controllers/trajectory_interpolation_controller.py

```diff
@@ -0,0 +1,122 @@
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+import math
+import numpy as np
+
+from smarts.core.coordinates import Heading, Pose
+from smarts.core.chassis import AckermannChassis
+
+from smarts.core.controllers.trajectory_tracking_controller import (
+    TrajectoryTrackingControllerState,
+    TrajectoryTrackingController,
+)
+
+TIME_INDEX = 0
+X_INDEX = 1
+Y_INDEX = 2
+THETA_INDEX = 3
+VEL_INDEX = 4
+
+
+class TrajectoryInterpolationController:
+    @staticmethod
+    def ndarray_has_nan(ndarray: np.ndarray):
+        np.isnan(np.sum(ndarray))
+
+    @staticmethod
+    def interpolate(ms0: np.ndarray, ms1: np.ndarray, time: float) -> np.ndarray:
+        """Linear Interpolate between two vehicle motion state
+
+        Returns:
+            np.ndarray: New vehicle state between vehicle motion state ms0 and ms1
+        """
+
+        start_time = ms0[TIME_INDEX]
+        end_time = ms1[TIME_INDEX]
+        ratio = math.fabs((time - start_time) / (end_time - start_time))
+        assert end_time >= start_time and time >= start_time
+
+        np_motion_state = (1 - ratio) * ms0 + ratio * ms1
+        return np_motion_state
+
+    @staticmethod
+    def locate_motion_state(trajectory, time) -> np.ndarray:
+
+        if math.isinf(trajectory[TIME_INDEX][0]):
+            return trajectory[:, 0], trajectory[:, 0]
+
+        end_index = 0
+        for i, t in enumerate(trajectory[TIME_INDEX]):
+            if t > time:
+                end_index = i
+                break
+
+        if end_index <= 0:
+            raise RuntimeError(
+                "Next step time is %f, but cant be located at input traj" % time
+            )
+
+        return trajectory[:, end_index - 1], trajectory[:, end_index]
+
+    @classmethod
+    def perform_trajectory_interpolation(
+        cls, sim, agent_id, vehicle, trajectory: np.ndarray, controller_state,
+    ):
+        """Move vehicle with trajectory interpolation
+
+        If you want vehicle stop at a specific pose,
+        trajectory[TIME_INDEX][0] should be setted as infi
```

Comment:
Sorry, I am confused. Do you mean "should be set as infinite"?

Reply:
Yes. Sorry for the typo. I'll fix it.

Reply:
@liamchzh I fixed it up with a force push.